RFC0003-Query-Dependencies.md

```diff
@@ -0,0 +1,119 @@
+# Summary
+
+Allow to specify a dependency on a query group without making it a super trait.
+
+# Motivation
+
+Currently, there's only one way to express that queries from group `A` can use
+another group `B`: namely, `B` can be a super-trait of `A`:
+
+```rust
+#[salsa::query_group(AStorage)]
+trait A: B {
+
+}
+```
+
+This approach works and allows one to express complex dependencies. However,
+this approach falls down when one wants to make a dependency a private
+implementation detail: Clients with `db: &impl A` can freely call `B` methods on
+the `db`.
+
+This is a bad situation from software engineering point of view: if everything
+is accessible, it's hard to make distinction between public API and private
+implementation details. In the context of salsa the situation is even worse,
+because it breaks "firewall" pattern. It's customary to wrap low-level
+frequently-changing or volatile queries into higher-level queries which produce
+stable results and contain invalidation. In the current salsa, however, it's
+very easy to accidentally call a low-level volatile query instead of a wrapper,
+introducing and undesired dependency.
+
+# User's guide
+
+To specify query dependencies, a `requires` clause should be used:
+
+```rust
+#[salsa::query_group(SymbolsDatabaseStorage, requires = "SyntaxDatabase + EnvDatabase")]
+pub trait SymbolsDatabase {
+    fn get_symbol_by_name(&self, name: String) -> Symbol;
+}
+```
+
+The value of `requires` is a `+` separated list of traits, like the one you would
+write in a `where` clause. These traits are available when implementing the query:
+
+```rust
+fn get_symbol_by_name(
+    db: &impl SymbolsDatabase + SyntaxDatabase + EnvDatabase,
```

Comment:
AFAIK, the compiler required parens here `&(impl A + B + C)`,
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=7d81641234462bb2a2f6daa55afadd98

Reply:
Thanks, fixed!